tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/auto_query_fixer/QueryTokenService.java

```diff
@@ -0,0 +1,85 @@
+package com.google.cloud.bigquery.utils.auto_query_fixer;
+
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.parser.babel.SqlBabelParserImpl;
+import org.apache.calcite.sql.parser.babel.Token;
+import org.apache.calcite.util.SourceStringReader;
+import org.apache.commons.lang3.tuple.Pair;
+
+import com.google.cloud.bigquery.utils.auto_query_fixer.entity.IToken;
+import com.google.cloud.bigquery.utils.auto_query_fixer.entity.TokenImpl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryTokenService {
+
+  public Pair<IToken, IToken> getNearbyTokens(String sql, int beginLine, int beginColumn) {
+    IToken previous = null;
+
+    for (IToken token : getAllTokens(sql)) {
+      if (token.getBeginLine() >= beginLine && token.getBeginCol() >= beginColumn) {
+        return Pair.of(previous, token);
+      }
+      previous = token;
+    }
+
+    return Pair.of(previous, previous);
+  }
+
+  public List<IToken> getAllTokens(String sql) {
+    List<IToken> tokens = new ArrayList<>();
+
+    // a parser contains token manager, which will be used to tokenize sql.
+    final SqlParser.ConfigBuilder configBuilder =
```

Comment:
Should this be a class member iso it can be used in other places too?

Reply:
Move this to the new class `BigQueryParserFactory`